From fdafbffd2be929d6eeeee1d2aff36e0141ca1983 Mon Sep 17 00:00:00 2001
From: Marc H. Thoben <napkin@develz.org>
Date: Wed, 21 Jan 2009 23:32:34 +0100
Subject: [PATCH] Adds !leave command
  * like kick but without temporary ban
  * new cvar g_adminLeaveMsg
  * using existing flag k

---
 src/game/g_admin.c |   54 ++++++++++++++++++++++++++++++++++++++++++++++++++++
 src/game/g_admin.h |    1 +
 src/game/g_local.h |    1 +
 src/game/g_main.c  |    2 +
 4 files changed, 58 insertions(+), 0 deletions(-)

diff --git a/src/game/g_admin.c b/src/game/g_admin.c
index a06bc7e..dffb7f6 100644
--- a/src/game/g_admin.c
+++ b/src/game/g_admin.c
@@ -139,6 +139,11 @@ g_admin_cmd_t g_admin_cmds[ ] =
       "save a map layout",
       "[^3mapname^7]"
     },
+
+    {"leave", G_admin_leave, "k",
+      "makes a player leave",
+      "[^3name|slot#^7] (^5additional reason^7)"
+    },
     
     {"listadmins", G_admin_listadmins, "D",
       "display a list of all server admins and their levels",
@@ -2012,6 +2017,55 @@ qboolean G_admin_kick( gentity_t *ent, int skiparg )
   return qtrue;
 }
 
+qboolean G_admin_leave( gentity_t *ent, int skiparg )
+{
+	int pids[ MAX_CLIENTS ];
+	char name[ MAX_NAME_LENGTH ], *reason, err[ MAX_STRING_CHARS ];
+	char notice[51];
+	char msg[51];
+
+	trap_Cvar_VariableStringBuffer( "g_banNotice", notice, sizeof( notice ) );
+	trap_Cvar_VariableStringBuffer( "g_adminLeaveMsg", msg, sizeof( msg ) );
+
+	if( G_SayArgc() < 2 + skiparg )
+	{
+		ADMP( "^3!leave: ^7usage: !leave [^3name|slot#^7] (^5additional reason^7)\n" );
+		return qfalse;
+	}
+	G_SayArgv( 1 + skiparg, name, sizeof( name ) );
+	reason = G_SayConcatArgs( 2 + skiparg );
+	if( G_ClientNumbersFromString( name, pids ) != 1 )
+	{
+		G_MatchOnePlayer( pids, err, sizeof( err ) );
+		ADMP( va( "^3!leave: ^7%s\n", err ) );
+		return qfalse;
+	}
+	if( !admin_higher( ent, &g_entities[ pids[ 0 ] ] ) )
+	{
+		ADMP( "^3!leave: ^7sorry, but your intended victim has a higher admin"
+			 " level than you\n" );
+		return qfalse;
+	}
+
+	trap_SendServerCommand( pids[ 0 ],
+						   va( "disconnect \"You have been forced to leave by %s^7\nreason:\n%s%s%s\n%s\"",
+							  ( ent ) ? va( "admin:\n%s", G_admin_adminPrintName( ent ) ) : "console",
+							  ( *msg ) ? msg : "not welcome",
+							  ( *reason ) ? ", " : "",
+                              ( *reason ) ? reason : "",
+							  notice
+							  ) );
+
+	trap_DropClient( pids[ 0 ], va( "was forced to leave%s^7, reason: %s%s%s",
+								   ( ent ) ? va( " by %s", G_admin_adminPrintName( ent ) ) : "console",
+								   ( *msg ) ? msg : "not welcome",
+								   ( *reason ) ? ", " : "",
+								   ( *reason ) ? reason : ""
+								   ) );
+
+	return qtrue;
+}
+
 qboolean G_admin_bot( gentity_t *ent, int skiparg ) {
 	// add [name] (team) (skill)
 	// del [name]
diff --git a/src/game/g_admin.h b/src/game/g_admin.h
index b88cb76..021b74e 100644
--- a/src/game/g_admin.h
+++ b/src/game/g_admin.h
@@ -204,6 +204,7 @@ qboolean G_admin_putmespec( gentity_t *ent, int skiparg );
 qboolean G_admin_warn( gentity_t *ent, int skiparg );
 qboolean G_admin_designate( gentity_t *ent, int skiparg );
 qboolean G_admin_cp( gentity_t *ent, int skiparg );
+qboolean G_admin_leave( gentity_t *ent, int skiparg );
 
 void G_admin_print( gentity_t *ent, char *m );
 void G_admin_buffer_print( gentity_t *ent, char *m );
diff --git a/src/game/g_local.h b/src/game/g_local.h
index 7890b73..4aae84b 100644
--- a/src/game/g_local.h
+++ b/src/game/g_local.h
@@ -1372,6 +1372,7 @@ extern  vmCvar_t  g_minLevelToJoinTeam;
 extern  vmCvar_t  g_forceAutoSelect;
 extern  vmCvar_t  g_minLevelToSpecMM1;
 extern  vmCvar_t  g_banNotice;
+extern  vmCvar_t  g_adminLeaveMsg;
 
 extern  vmCvar_t  g_devmapKillerHP;
 
diff --git a/src/game/g_main.c b/src/game/g_main.c
index 2bba217..66e90e1 100644
--- a/src/game/g_main.c
+++ b/src/game/g_main.c
@@ -163,6 +163,7 @@ vmCvar_t  g_publicSayadmins;
 vmCvar_t  g_myStats;
 vmCvar_t  g_antiSpawnBlock;
 vmCvar_t  g_banNotice;
+vmCvar_t  g_adminLeaveMsg;
 
 vmCvar_t  g_devmapKillerHP;
 
@@ -354,6 +355,7 @@ static cvarTable_t   gameCvarTable[ ] =
   { &g_rankings, "g_rankings", "0", 0, 0, qfalse },
   { &g_allowShare, "g_allowShare", "0", CVAR_ARCHIVE | CVAR_SERVERINFO, 0, qfalse},
   { &g_banNotice, "g_banNotice", "", CVAR_ARCHIVE, 0, qfalse  },
+  { &g_adminLeaveMsg, "g_adminLeaveMsg", "", CVAR_ARCHIVE, 0, qfalse  },
 };
 
 static int gameCvarTableSize = sizeof( gameCvarTable ) / sizeof( gameCvarTable[ 0 ] );
-- 
1.6.0.4

